src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,107 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private static final double GROW_MULTIPLIER = 1.5;
+    private int size;
+    private Object[] elementData;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_SIZE];
+
+    }
+
+    public void checkCapacity() {
+        if (size == elementData.length) {
+            int newSize = (int)(elementData.length * GROW_MULTIPLIER);
+            Object[] newArray = new Object[newSize];
+            System.arraycopy(elementData, 0, newArray, 0, size);
+            elementData = newArray;
+        }
+    }
+
     @Override
     public void add(T value) {
+        checkCapacity();
+        elementData[size++] = value;
 
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForSet(index);
+        checkCapacity();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
 
     }
 
+    public void checkIndexForSet(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+    }
```

Comment:
The method `checkIndexForSet` is redundant since it performs similar logic to the `checkIndex` method. Consider using a single method for index validation to avoid code repetition, as suggested in the checklist: "If the logic of your code repeats - move it to a separate private method."